### ISSUE TITLE
ExampleTableViewController - variables naming, misallocation

### ISSUE THREAD

**Issue by SuperWomble** [loosestar/MissTableViewNavigationController#8]
In `viewDidLoad()`:

```
let leftButton: UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonItem.SystemItem.add, target: self, action: #selector(ExampleTableViewController.handleNewButtonItemTapped))

        let rightButton: UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonItem.SystemItem.edit, target: nil, action: #selector(ExampleTableViewController.handleEditButtonItemTapped))
```

**1. NAMING**

'leftButton' and 'rightButton' are instances of `UIBarButtonItem`; should be named as such.

Further, they are local variables. The instances will remain alive as long as the NavigationBar holds on to them, but you have lost a reference to them. 

**2. BUG**

```
self.navigationItem.leftBarButtonItem = leftButton
self.navigationItem.rightBarButtonItem = self.editButtonItem
```
should be

`self.navigationItem.rightBarButtonItem = rightBarButtonItem`

**3. UNUSED INSTANCE VAR**

```
 var rightBarButtonItem: UIBarButtonItem {
        let barButtonItem = UIBarButtonItem(title: "Edit", style: .plain, target: self, action: nil)
        barButtonItem.tintColor = UIColor.blue
        return barButtonItem
    }
```

This is not being used (and the action is set to nil) but it should be used. See point (4).

**4. LOST REFERENCES**

When you create the `UIBarButtonItem` instances inside `viewDidLoad()`, they are being assigned to temporary references. Those references will be lost as soon as `viewDidLoad()` falls out of scope.

The instances will remain in memory as long as that navigationBar holds them, as it holds them with a strong reference.  

So you can get access to them through the `navigationItem`, as you are doing, but you could just use two instance vars instead, to hold strong references to the items. These will last as long as the viewController does.

**Comment by SuperWomble** [comment]
I made edits to the above. I made a mistake about using self.editButtonItem.

**Comment by loosestar** [comment]
Took half an hour of debugging (until today, rightButton's target was nil which of course compiled but did nothing with the action/handler), but I'm now using leftButton and rightButton. Will update further to fit with your private instance variable suggestion.

**Comment by SuperWomble** [comment]
The commit comment says that you are using  private instance variables, but the source code isn't show that.

Also, self.editButtonItem is not being used to create the Edit button.

` self.navigationItem.rightBarButtonItem = rightButton //self.editButtonItem`


### **Terminology Recap**

A **local** variable (or function or type) is one that belongs to a _specific scope_.

An **instance** variable (or function or type) is one that belongs to a specific _instance_ of the Struct/Class.

A **static** variable (or function or type) is one that belongs to a a Type (e.g. Struct, Class, Enum), and which is _shared_.

e.g.
```

struct Foo
{
    /// This is a static property. It belongs to Foo. There is only one of these in existence.
    static let kMagicScale: CGFloat = 3.14

    /// This is a static function. It belongs to Foo. It does not, cannot, use any instance
    /// instance variables.
    /// Static functions are often used as "maker" functions: conveniences to make 
    /// pre-configured instances of the type. Also known as "factories".

    static func makeBiggestFoo() -> Foo
    {
        return Foo(scale: .infinity)
    }
    
    /// This is an instance variable. Every instance of Food will have its own copy of this.
    var scale: CGFloat = 1

    /// This is an instance method. It can be called only by instances of Foo.
    func debugPrint()
    {
        /// This is a local variable. It ceases to exist once the function ends. i.e. it falls out of scope.
        let timeStamp = Date()
        
        print("debugging started at: \(timeStamp)"

        if self.scale.isInfinite {
            print("It's turtles all the way down!")
        } else {
            print("scale: \(self.scale)"
        }
    }
}

```

...```

let regularFoo = Foo(scale: 1)
regularFoo.debugPrint()

let magicFoo = Foo(scale: Foo.kMagicScale)

let biggestFoo = Foo.makeBiggestFoo()

```

**Comment by SuperWomble** [comment]
The problem you are running in to is that you want to keep references to objects (UIBarButtonItems) but can't do it at init type. 

And you **can't** do it at init time, because the 'self' is called on during init yet 'self' isn't fully set up until init() returns.

This is why lazy properties were invented. They are executed once, and only once, when first used.  The excution is thus deferred until safely after init() has completed. 

The use of `fileprivate` (or `private` or `open` or `public`) doesn't matter.

In the following code, note how the _addBarButtonItem var is being set. What we are doing there is creating a closure, and then executing that closure. This allows us to do complex assignment to the var.

Again, that closure will only be called once, when _addBarButtonItem property is first accessed (which for us in viewDidLoad)


e.g.

	```
	
	fileprivate lazy var _editBarButtonItem = self.editButtonItem
	
	fileprivate lazy var _addBarButtonItem: UIBarButtonItem = {
		return UIBarButtonItem(barButtonSystemItem: .add,
				target: self,
				action: #selector(handleAddButtonItemTapped))
	}()
	
	@objc func handleAddButtonItemTapped(sender: UIBarButtonItem)
	{
		// TODO: complete
		print("Add button tapped")
	}
	
	override func viewDidLoad()
	{
		super.viewDidLoad()
		
		self.view.backgroundColor = .lightGray
		self.navigationItem.leftBarButtonItem = _editBarButtonItem
		self.navigationItem.rightBarButtonItem = _addBarButtonItem
    }
	

```

Regarding the use of the Edit button, please read this:

https://developer.apple.com/documentation/uikit/uiviewcontroller/1621471-editbuttonitem

**Comment by loosestar** [comment]
Thanks for the pointer. I'm still reading up on the intended use of lazy variables (before making use of them and not being sure of what I'm doing), but this makes it much clearer.
As for editButtonItem...::facepalm::
